### ISSUE TITLE
Git branching model

### ISSUE THREAD

**Issue by DerTanzschuh** [git-ftp/git-ftp#517]
I would like to suggest to use a defined branching strategy.

At the moment there are various branches with different purposes. And the master branch is not in every case the most recent stable version of the tool.

By defining a branching strategy we could organize the development and release workflow and make sure the master branch is always stable. I don’t think it’s such a huge overhead but it brings many advantages.

**I can recommend the [Git branching model by Vincent Driessen](https://nvie.com/posts/a-successful-git-branching-model/).** It’s quite easy to handle it but powerful.

What do you think of that? Have you an other model we could use?

---

I would also be willing to check the branches not merged yet and maybe rebase them on top of a new `development` branch. We could use this point to kind of clean up the repository a bit and delete legacy branches that are merged or implemented in another branch already.

**Comment by mkllnk** [comment]
Thank you for thinking this through. What are your criteria to declare a version as stable?

This is a very small project and therefore I would like to keep things as simple as possible. I see this structure:

- `master` is the latest stable version. Since we don't have any beta testing program, there wouldn't be any difference between a `develop` and a `master` branch. We don't have any additional tests when transitioning to `master`.
- We mainly support only the latest release. There are still some older branches to support older release lines, e.g. `0.9.x`. Older ones can be removed. The oldest I would consider supporting is the version currently used in Debian stable. So if there is a bug in that older version that needs fixing, we can do that. But most bug reports are about the latest release and most people are comfortable switching to the latest release if they encountered a bug.
- We also have feature branches.

In summary, I see our model as the same as you suggested with the only difference that `master` and `develop` are the same.

I've been working on projects that keep merging pull requests for a few weeks and then to a big manual release testing round to declare something as stable and release it. That's a lot of effort. And I also worked in a team were we tested every pull requests as much as we could or was reasonable and then declared `master` as stable in general. We were even deploying that app multiple times a day, after each merge. That had the advantage of detecting errors early and looking for errors in a much smaller changeset. I really liked that and would suggest the same here. Just trying to develop *stable* code at all times. It makes even more sense when the team is so small and there are only a handful of merges per year. Then every merge is like a release, I just wouldn't go through the effort of actually creating a release for every merge. That's too much work for me. ;-)

**Comment by DerTanzschuh** [comment]
I should've said better “most recent **released** version” .

> In summary, I see our model as the same as you suggested with the only difference that master and develop are the same.

I agree the difference isn't that huge. The advantage of separating the master and the develop branch would be, that the master is always a released version – at least for me this is a advantage. The develop branch would be the unreleased version. Of course It is good to ensure it is stable (in the sense of running without issues and passing all the tests) as well to target the problems you noted.

Why is this a advantage? It hides the current (unreleased) version for users passing by. So if somebody comes by and there is a description for a new feature in the ReadMe, he might think it’s working in the current version. By defining master as released version this would fix that. 

If a separated development branch is used a release would be like:
- bump version in develop / release-X
- add changes to change log
- merge to master
- tag master

If think its quite the same effort to do a release as today (only the merge to master is new) but with a master branch that always is a released version.

**TL;DR** 
But as you noted and as I said, the difference isn’t that huge. The key difference is that the master is always a released version and to separate the development branch just for organisational purposes.

But maybe this is a bit of a overkill for this project. I just wanted to mention the idea.

**Comment by mkllnk** [comment]
Great input. I see the advantage of showing documentation on Github for the same version that people have installed locally. In reality, there will always be many different version around, some newer and some older. So even if we had master pointing to the last release, there would still be people getting confused. I think that we should improve the documentation to mention which version support certain features. I haven't really done that in the past but I would like to start with that now. So if people read about a feature then they know which version supports it and they can check their own version.

So, I think I would like to keep the current branching model. But do you think we should document it? You are probably not the only one wondering about this. Maybe it would help people who would like to support the project.

**Comment by DerTanzschuh** [comment]
> I think that we should improve the documentation to mention which version support certain features.

Isn't that the purpose of the changelog?

But I agree, it is a good idea to improve the documentation to make the changeling more prominent. (There is no link to it in the manual or the ReadMe.)

I would say it is enough to add a note at the beginning and say something like:
> This is the manual for version XX. Please consider the changelog or select your version from the _branch > Tags_ above to see the manual for this version.

Of course, it must not be forgotten to change this with every release. But I'd find it rather confusing if there would be version information scattered around all over in the manual.

**Comment by mkllnk** [comment]
> I'd find it rather confusing if there would be version information scattered around all over in the manual.

Oh, really? Let's take an example. What do you think about this one?

```diff
diff --git a/man/git-ftp.1.md b/man/git-ftp.1.md
index ede84b2..505a1bb 100644
--- a/man/git-ftp.1.md
+++ b/man/git-ftp.1.md
@@ -172,7 +172,7 @@ different and handles only those files. That saves time and bandwidth.
 :      Fails if post-ftp-push raises an error.
 
 `--auto-init`
-:      Automatically run init action when running push action
+:      Automatically run init action when running push action. This option was introduced in version 1.3.3.
 
 `--version`
 :      Prints version.
```

**Comment by DerTanzschuh** [comment]
Yes, this isn’t to bad, but on the one hand this information can be found in the changelog easily and on the other hand – what if I wan’t to not what is supported by a specific version? It’s easier to say: „just consult the manual of the version you are using“.

I think this is covered by https://github.com/git-ftp/git-ftp/pull/523/commits/2c7f408fd285950b4454d824c339d798dfcabf8f and merged with #523.

---

So we decided to stay with the current branches, since this model is enough for this project. 🙂  I’ll check some of the older branches if I get the time.
